File: haskell/Readline.hs
module Readline
( readline, load_history )
where

-- Pick one of these:
-- GPL license
import qualified System.Console.Readline as RL
-- BSD license
--import qualified System.Console.Editline.Readline as RL

import Control.Monad (when)
import System.Directory (getHomeDirectory, doesFileExist)
import System.IO.Error (tryIOError)

history_file :: IO String
history_file = do
    home <- getHomeDirectory
    return $ home ++ "/.mal-history"

load_history :: IO ()
load_history = do
    hfile <- history_file
    fileExists <- doesFileExist hfile
    when fileExists $ do
        content <- readFile hfile
        mapM_ RL.addHistory (lines content)

readline :: String -> IO (Maybe String)
readline prompt = do
    maybeLine <- RL.readline prompt
    case maybeLine of
        Nothing -> return ()
        Just "" -> return ()
        Just line@(_:_) -> do
            hfile <- history_file
            _ <- tryIOError (appendFile hfile (line ++ "\n"))
            RL.addHistory line
    return maybeLine

File: clojure/src/readline.clj
(ns readline
    (:require [clojure.string :refer [split]]
              [net.n01se.clojure-jna :as jna]))

(defonce history-loaded (atom nil))
(def HISTORY-FILE (str (System/getProperty "user.home") "/.mal-history"))

;;
;; Uncomment one of the following readline libraries
;;

;; editline (BSD)
#_
(do
  (def readline-call (jna/to-fn String edit/readline))
  (def add-history (jna/to-fn Void edit/add_history))
  (def load-history #(doseq [line (split (slurp %) #"\n")]
                       (jna/invoke Void edit/add_history line))))

;; GNU Readline (GPL)
;; WARNING: distributing your code with GNU readline enabled means you
;; must release your program as GPL
;#_
(do 
  (def readline-call (jna/to-fn String readline/readline))
  (def add-history (jna/to-fn Void readline/add_history))
  (def load-history (jna/to-fn Integer readline/read_history)))

(defn readline [prompt & [lib]]
  (when (not @history-loaded)
    (reset! history-loaded true)
    (load-history HISTORY-FILE))
  (let [line (readline-call prompt)]
    (when line
      (add-history line)
      (spit HISTORY-FILE (str line "\n") :append true))
    line))
